src/openms/source/ANALYSIS/RNPXL/HyperScore.cpp

```diff
@@ -0,0 +1,115 @@
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Timo Sachsenberg $
+// $Authors: Timo Sachsenberg $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/ANALYSIS/RNPXL/HyperScore.h>
+
+#include <vector>
+#include <map>
+
+using std::vector;
+
+namespace OpenMS
+{
+  double HyperScore::logfactorial_(UInt x)
+  {
+    UInt y;
+
+    if (x < 2)
+      return 1;
+    else
+    {
+      double z = 0;
+      for (y = 2; y <= x; y++)
+      {
+        z = log((double)y) + z;
+      }
+
+      return z;
+    }
+  }
+
+  double HyperScore::compute(double fragment_mass_tolerance, bool fragment_mass_tolerance_unit_ppm, const PeakSpectrum& exp_spectrum, const RichPeakSpectrum& theo_spectrum)
+  {
+    double dot_product = 0.0;
+    UInt y_ion_count = 0;
+    UInt b_ion_count = 0;
+
+    for (MSSpectrum<RichPeak1D>::ConstIterator theo_peak_it = theo_spectrum.begin(); theo_peak_it != theo_spectrum.end(); ++theo_peak_it)
+    {
+      const double& theo_mz = theo_peak_it->getMZ();
+
+      double max_dist_dalton = fragment_mass_tolerance_unit_ppm ? theo_mz * fragment_mass_tolerance * 1e-6 : fragment_mass_tolerance;
+
+      // iterate over peaks in experimental spectrum in given fragment tolerance around theoretical peak
+      Size index = exp_spectrum.findNearest(theo_mz);
+      double exp_mz = exp_spectrum[index].getMZ();
+
+      // found peak match
+      if (std::abs(theo_mz - exp_mz) < max_dist_dalton)
+      {
+        dot_product += exp_spectrum[index].getIntensity();
```

Comment:
well if it is actually a dotproduct, then that would make sense :-) 

I think this would make it possible to set for example all y ions to intensity 1 and all b ions to intensity 0.5 (and neutral losses again to a different one) depending on the instrument. I think some search engines do this. So to me it would increase the possibilities and utility of the algo